File: go/valid_palindrome_ii.go
package leetcode

// Time complexity: O(n)
// Space complexity: O(1)
func validPalindrome(s string) bool {

	return check(0, len(s)-1, s, false)
}

func check(left, right int, s string, found bool) bool {
	for left < right {
		if s[left] != s[right] {
			if found == true {
				return false
			} else {
				return check(left+1, right, s, true) || check(left, right-1, s, true)
			}
		} else {
			left++
			right--
		}
	}
	return true
}

File: python/valid-palindrome-ii.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(1)
    def validPalindrome(self, s: str) -> bool:
        def check(left, right, found=False):
            while left < right:
                if s[left] == s[right]:
                    left, right = left+1, right-1
                elif found == False:
                    return check(left+1, right, True) or check(left, right-1, True)
                else:
                    return False
            return True

        return check(0, len(s)-1)
